compiler/src/iree/compiler/InputConversion/TOSA/TosaToLinalgExt.cpp

```diff
@@ -0,0 +1,187 @@
+//===- TosaToLinalgExt.cpp --------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Convert TOSA operations to their LinalgExt equivalent.
+//
+//===----------------------------------------------------------------------===//
+
+#include "iree-dialects/Dialect/LinalgExt/IR/LinalgExtDialect.h"
+#include "iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.h"
+#include "iree/compiler/InputConversion/TOSA/PassDetail.h"
+#include "iree/compiler/InputConversion/TOSA/Passes.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/Dialect/Tosa/Transforms/Passes.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace mlir;
+using namespace mlir::tosa;
+
+namespace mlir {
+namespace iree_compiler {
+
+static SmallVector<utils::IteratorType> getNParallelLoopsAttrs(
+    unsigned nParallelLoops) {
+  return SmallVector<utils::IteratorType>(nParallelLoops,
+                                          utils::IteratorType::parallel);
+}
+
+// Converts tosa.scatter to the iree_linalg_ext.scatter operation. As the
+// LinalgExt version is not batched therefore we materialize the batch index
+// for each update.
+class ScatterConversion : public OpRewritePattern<tosa::ScatterOp> {
+ public:
+  using OpRewritePattern<tosa::ScatterOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tosa::ScatterOp op,
+                                PatternRewriter &rewriter) const final {
+    auto values = op.getValuesIn();
+    auto indices = op.getIndices();
+    auto updates = op.getInput();
+    auto valuesTy = values.getType().cast<ShapedType>();
+    auto indicesTy = indices.getType().cast<ShapedType>();
+    auto updatesTy = updates.getType().cast<ShapedType>();
+    ImplicitLocOpBuilder builder(op.getLoc(), rewriter);
+
+    if (!valuesTy.hasRank() || !indicesTy.getRank() || !updatesTy.getRank())
```

Comment:
Did you mean to dyn_cast<RankedTensorType> above? (with the new support LLVM added we can now do `dyn_cast<Foo>(bla.getType())` and be more uniform than we could before).

Reply:
I switched toe a dyn_cast to `RankedTensorType` and ommitted the cast. I left as `<foo>.dyn_cast<BAR>()`, if there is a strong preference for otherwise but I thought that format was more common throughout the codebase. If it isn't I can change.

Reply:
One is more common for operations & attributes, the other for types, I'd like to have it all use 1 style, but not blocking.